gtsfm/averaging/rotation/cycle_consistency.py

```diff
@@ -0,0 +1,234 @@
+"""
+A library for cycle triplet extraction and cycle error computation.
+
+Checks the cumulative rotation errors between triplets to throw away cameras.
+Note: the same property does not hold for cumulative translation errors when scale is unknown (i.e. in SfM).
+
+Author: John Lambert
+"""
+
+import os
+from collections import defaultdict
+from typing import Dict, List, Optional, Tuple
+
+import matplotlib.pyplot as plt
+import numpy as np
+from gtsam import Rot3, Unit3
+
+import gtsfm.utils.geometry_comparisons as comp_utils
+import gtsfm.utils.logger as logger_utils
+from gtsfm.two_view_estimator import TwoViewEstimationReport
+
+
+logger = logger_utils.get_logger()
+
+
+CYCLE_ERROR_THRESHOLD = 5.0
+
+
+def extract_triplets(i2Ri1_dict: Dict[Tuple[int, int], Rot3]) -> List[Tuple[int, int, int]]:
+    """Discover triplets from a graph, without O(n^3) complexity, by using intersection within adjacency lists.
+
+    Based off of Theia's implementation:
+        https://github.com/sweeneychris/TheiaSfM/blob/master/src/theia/math/graph/triplet_extractor.h
+
+    If we have an edge a<->b, if we can find any node c such that a<->c and b<->c, then we have
+    discovered a triplet. In other words, we need only look at the intersection between the nodes
+    connected to `a` and the nodes connected to `b`.
+
+    Args:
+        i2Ri1_dict: mapping from image pair indices to relative rotation.
+
+    Returns:
+        triplets: 3-tuples of nodes that form a cycle. Nodes of each triplet are provided in sorted order.
+    """
+    # only want to keep the unique ones
+    triplets = set()
+
+    # form adjacency list
+    adj_list = defaultdict(set)
+
+    for (i1, i2), i2Ri1 in i2Ri1_dict.items():
+        if i2Ri1 is None:
+            continue
+
+        adj_list[i1].add(i2)
+        adj_list[i2].add(i1)
+
+    # find intersections
+    for (i1, i2), i2Ri1 in i2Ri1_dict.items():
```

Comment:
The errors accumulate as the cycle length increases. I think there are papers that don’t recommend using larger cycles for outlier rejection. We would end up rejecting more edges than what we added (by increasing look-ahead window). If we really want to do that we should weigh the errors by the inverse of their cycle length, which increases hand-tuned parameters. Shonan is also quite robust already I think. 

I agree we must use more edges than the hard-coded look ahead window, but even if we do, it may not be worth using cycles with length > 3 for outlier rejection.

Reply:
I wasn't saying to increase the cycle length; the cycle length would still be three. I'm saying to check _all_ of the valid triplets cycles in the graph based on matches between the frames, not just the ones within the look ahead window.